File: packages/swift/Sources/ReallyMeCrypto/JwaConcatKdf.swift
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import CryptoKit
import Foundation

/// JWA ECDH-ES Concat KDF over SHA-256.
///
/// This helper owns only the cryptographic derivation. Compact JWE parsing,
/// header policy, `alg`/`enc` validation, and payload decoding belong in the
/// JOSE layer that calls it.
public enum ReallyMeJwaConcatKdf {
    public static let sha256DigestLength = 32
    public static let maxSharedSecretLength = 4096
    public static let maxInfoLength = 4096
    public static let minOutputLength = 1
    public static let maxOutputLength = 4096

    public static func deriveSha256(
        sharedSecret: [UInt8],
        algorithmId: [UInt8],
        partyUInfo: [UInt8],
        partyVInfo: [UInt8],
        outputLength: Int
    ) throws -> [UInt8] {
        try validate(
            sharedSecret: sharedSecret,
            algorithmId: algorithmId,
            partyUInfo: partyUInfo,
            partyVInfo: partyVInfo,
            outputLength: outputLength
        )

        let outputBits = outputLength * 8
        var otherInfo = try buildOtherInfo(
            algorithmId: algorithmId,
            partyUInfo: partyUInfo,
            partyVInfo: partyVInfo,
            outputBits: UInt32(outputBits)
        )
        let reps = (outputLength + sha256DigestLength - 1) / sha256DigestLength
        var derived = [UInt8]()
        derived.reserveCapacity(reps * sha256DigestLength)

        for counter in 1...reps {
            var hasher = SHA256()
            hasher.update(data: Data(uint32be(UInt32(counter))))
            hasher.update(data: Data(sharedSecret))
            hasher.update(data: Data(otherInfo))
            derived.append(contentsOf: hasher.finalize())
        }

        let output = Array(derived.prefix(outputLength))
        ReallyMeCryptoMemory.bestEffortClear(&derived)
        ReallyMeCryptoMemory.bestEffortClear(&otherInfo)
        return output
    }

    private static func validate(
        sharedSecret: [UInt8],
        algorithmId: [UInt8],
        partyUInfo: [UInt8],
        partyVInfo: [UInt8],
        outputLength: Int
    ) throws {
        guard !sharedSecret.isEmpty,
              sharedSecret.count <= maxSharedSecretLength,
              !algorithmId.isEmpty,
              algorithmId.count <= maxInfoLength,
              partyUInfo.count <= maxInfoLength,
              partyVInfo.count <= maxInfoLength,
              (minOutputLength...maxOutputLength).contains(outputLength)
        else {
            throw ReallyMeCryptoError.invalidInput
        }
    }

    private static func buildOtherInfo(
        algorithmId: [UInt8],
        partyUInfo: [UInt8],
        partyVInfo: [UInt8],
        outputBits: UInt32
    ) throws -> [UInt8] {
        var otherInfo = [UInt8]()
        otherInfo.reserveCapacity(
            lengthPrefixedCapacity(algorithmId)
                + lengthPrefixedCapacity(partyUInfo)
                + lengthPrefixedCapacity(partyVInfo)
                + 4
        )
        try appendLengthPrefixed(algorithmId, to: &otherInfo)
        try appendLengthPrefixed(partyUInfo, to: &otherInfo)
        try appendLengthPrefixed(partyVInfo, to: &otherInfo)
        otherInfo.append(contentsOf: uint32be(outputBits))
        return otherInfo
    }

    private static func lengthPrefixedCapacity(_ bytes: [UInt8]) -> Int {
        4 + bytes.count
    }

    private static func appendLengthPrefixed(_ bytes: [UInt8], to output: inout [UInt8]) throws {
        guard let length = UInt32(exactly: bytes.count) else {
            throw ReallyMeCryptoError.invalidInput
        }
        output.append(contentsOf: uint32be(length))
        output.append(contentsOf: bytes)
    }

    private static func uint32be(_ value: UInt32) -> [UInt8] {
        [
            UInt8((value >> 24) & 0xff),
            UInt8((value >> 16) & 0xff),
            UInt8((value >> 8) & 0xff),
            UInt8(value & 0xff),
        ]
    }
}

File: packages/kotlin/src/main/kotlin/me/really/crypto/JwaConcatKdf.kt
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

package me.really.crypto

import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

/**
 * JWA ECDH-ES Concat KDF over SHA-256.
 *
 * JOSE parsing and header policy stay outside this primitive; callers pass
 * already-decoded `AlgorithmID`, `PartyUInfo`, and `PartyVInfo` values.
 */
public object ReallyMeJwaConcatKdf {
    public const val SHA256_DIGEST_LENGTH: Int = 32
    public const val MAX_SHARED_SECRET_LENGTH: Int = 4096
    public const val MAX_INFO_LENGTH: Int = 4096
    public const val MIN_OUTPUT_LENGTH: Int = 1
    public const val MAX_OUTPUT_LENGTH: Int = 4096

    public fun deriveSha256(
        sharedSecret: ByteArray,
        algorithmId: ByteArray,
        partyUInfo: ByteArray,
        partyVInfo: ByteArray,
        outputLength: Int,
    ): ByteArray {
        validate(sharedSecret, algorithmId, partyUInfo, partyVInfo, outputLength)
        val outputBits = checkedOutputBits(outputLength)
        val otherInfo = buildOtherInfo(algorithmId, partyUInfo, partyVInfo, outputBits)
        val reps = (outputLength + SHA256_DIGEST_LENGTH - 1) / SHA256_DIGEST_LENGTH
        val derived = ByteArray(reps * SHA256_DIGEST_LENGTH)

        for (counter in 1..reps) {
            val digest = sha256Digest()
            val counterBytes = uint32be(counter)
            digest.update(counterBytes)
            digest.update(sharedSecret)
            digest.update(otherInfo)
            val block = digest.digest()
            block.copyInto(derived, destinationOffset = (counter - 1) * SHA256_DIGEST_LENGTH)
            block.fill(0)
            counterBytes.fill(0)
        }

        val output = derived.copyOf(outputLength)
        derived.fill(0)
        otherInfo.fill(0)
        return output
    }

    private fun validate(
        sharedSecret: ByteArray,
        algorithmId: ByteArray,
        partyUInfo: ByteArray,
        partyVInfo: ByteArray,
        outputLength: Int,
    ) {
        if (sharedSecret.isEmpty() ||
            sharedSecret.size > MAX_SHARED_SECRET_LENGTH ||
            algorithmId.isEmpty() ||
            algorithmId.size > MAX_INFO_LENGTH ||
            partyUInfo.size > MAX_INFO_LENGTH ||
            partyVInfo.size > MAX_INFO_LENGTH ||
            outputLength !in MIN_OUTPUT_LENGTH..MAX_OUTPUT_LENGTH
        ) {
            throw ReallyMeCryptoException.InvalidInput()
        }
    }

    private fun checkedOutputBits(outputLength: Int): Int {
        val outputBits = outputLength.toLong() * Byte.SIZE_BITS.toLong()
        if (outputBits > Int.MAX_VALUE.toLong()) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        return outputBits.toInt()
    }

    private fun buildOtherInfo(
        algorithmId: ByteArray,
        partyUInfo: ByteArray,
        partyVInfo: ByteArray,
        outputBits: Int,
    ): ByteArray {
        val totalLength =
            lengthPrefixedCapacity(algorithmId) +
                lengthPrefixedCapacity(partyUInfo) +
                lengthPrefixedCapacity(partyVInfo) +
                UINT32_LENGTH
        val output = ByteArray(totalLength)
        var offset = appendLengthPrefixed(output, 0, algorithmId)
        offset = appendLengthPrefixed(output, offset, partyUInfo)
        offset = appendLengthPrefixed(output, offset, partyVInfo)
        uint32be(outputBits).copyInto(output, destinationOffset = offset)
        return output
    }

    private fun lengthPrefixedCapacity(bytes: ByteArray): Int = UINT32_LENGTH + bytes.size

    private fun appendLengthPrefixed(output: ByteArray, offset: Int, bytes: ByteArray): Int {
        uint32be(bytes.size).copyInto(output, destinationOffset = offset)
        val valueOffset = offset + UINT32_LENGTH
        bytes.copyInto(output, destinationOffset = valueOffset)
        return valueOffset + bytes.size
    }

    private fun uint32be(value: Int): ByteArray =
        byteArrayOf(
            ((value ushr 24) and 0xff).toByte(),
            ((value ushr 16) and 0xff).toByte(),
            ((value ushr 8) and 0xff).toByte(),
            (value and 0xff).toByte(),
        )

    private fun sha256Digest(): MessageDigest =
        try {
            MessageDigest.getInstance("SHA-256")
        } catch (_: NoSuchAlgorithmException) {
            throw ReallyMeCryptoException.ProviderFailure()
        }

    private const val UINT32_LENGTH = 4
}
